sharding/networksharding/peerShardMapper.go

```diff
@@ -76,74 +76,140 @@ func NewPeerShardMapper(
 	}, nil
 }
 
-// GetShardID returns the corresponding shard ID of a given peer ID.
-// If the pid is unknown, it will return sharding.UnknownShardId
-func (psm *PeerShardMapper) GetShardID(pid p2p.PeerID) (shardId uint32) {
-	shardId, pk, ok := psm.getShardIDWithNodesCoordinator(pid)
+// GetPeerInfo returns the corresponding shard ID of a given peer ID.
+// It also return the type of provided peer
+func (psm *PeerShardMapper) GetPeerInfo(pid p2p.PeerID) core.P2PPeerInfo {
+	pInfo, pk, ok := psm.getPeerInfoWithNodesCoordinator(pid)
 	if ok {
-		return shardId
+		return pInfo
 	}
 
-	shardId, ok = psm.getShardIDSearchingPkInFallbackCache(pk)
+	shardId, ok := psm.getShardIDSearchingPkInFallbackCache(pk)
 	if ok {
-		return shardId
+		return core.P2PPeerInfo{
+			PeerType: core.ObserverdPeer,
+			ShardID:  shardId,
+		}
 	}
 
-	return psm.getShardIDSearchingPidInFallbackCache(pid)
+	return psm.getPeerInfoSearchingPidInFallbackCache(pid)
 }
 
-func (psm *PeerShardMapper) getShardIDWithNodesCoordinator(pid p2p.PeerID) (shardId uint32, pk []byte, ok bool) {
+func (psm *PeerShardMapper) getPeerInfoWithNodesCoordinator(pid p2p.PeerID) (peerInfo core.P2PPeerInfo, pk []byte, ok bool) {
 	pkObj, ok := psm.peerIdPk.Get([]byte(pid))
 	if !ok {
-		return core.UnknownShardId, nil, false
+		log.Trace("unknown peer",
+			"pid", p2p.PeerIdToShortString(pid),
+			"pk", pk,
+		)
+
+		return core.P2PPeerInfo{
+			PeerType: core.UnknownPeer,
+			ShardID:  0,
+		}, nil, false
 	}
 
 	pkBuff, ok := pkObj.([]byte)
 	if !ok {
 		log.Warn("PeerShardMapper.getShardIDWithNodesCoordinator: the contained element should have been of type []byte")
-		return core.UnknownShardId, nil, false
+
+		return core.P2PPeerInfo{
+			PeerType: core.UnknownPeer,
+			ShardID:  0,
+		}, nil, false
 	}
 
 	_, shardId, err := psm.nodesCoordinator.GetValidatorWithPublicKey(pkBuff, psm.epochHandler.Epoch())
 	if err != nil {
-		return core.UnknownShardId, pkBuff, false
+		log.Trace("unknown peer",
+			"pid", p2p.PeerIdToShortString(pid),
+			"pk", pk,
+			"error", err,
+		)
+
+		return core.P2PPeerInfo{
+			PeerType: core.UnknownPeer,
+			ShardID:  0,
+		}, pkBuff, false
 	}
 
-	return shardId, pkBuff, true
+	log.Trace("peer",
+		"pid", p2p.PeerIdToShortString(pid),
+		"pk", pk,
```

Comment:
The display of pk will be correct?

Reply:
Yes, as pk is of type []byte (automatically hexed)